File: go/leet_code/88.go
// Package leet_code provides solutions for LeetCode problems.
//
// This file provides a solution for the problem of merging two sorted arrays.
package leet_code

// merge merges two sorted arrays in-place.
//
// It accepts four parameters:
// - nums1: a slice of integers representing the first sorted array.
// - m: an integer representing the number of initialized elements in nums1.
// - nums2: a slice of integers representing the second sorted array.
// - n: an integer representing the number of initialized elements in nums2.
//
// The function does not return a value. It modifies nums1 in-place to include the elements of nums2 in sorted order.
//
// The function iterates from the end of nums1, comparing the last elements of nums1 and nums2, and placing the larger one at the current position.
// If all elements in nums1 are processed (m == 0), the remaining elements in nums2 are copied to nums1.
// If all elements in nums2 are processed (n == 0), the function returns as the remaining elements in nums1 are already in place.
//
// Time complexity analysis:
// - Best-case: O(n), when all elements in nums1 are larger than those in nums2.
// - Worst-case: O(n), when all elements in nums2 are larger than those in nums1.
// - Average-case: O(n), as we always have to iterate over each element in nums1 and nums2.
func merge(nums1 []int, m int, nums2 []int, n int) {
	for i := m + n - 1; i >= 0; i-- {
		if m == 0 {
			nums1[i] = nums2[n-1]
			n--
			continue
		}

		if n == 0 {
			return
		}

		if nums1[m-1] > nums2[n-1] {
			nums1[i] = nums1[m-1]
			m--
		} else {
			nums1[i] = nums2[n-1]
			n--
		}
	}
}

File: dart/leet_code/88.dart
/// Dart implementation of the LeetCode problem 88: Merge Sorted Array.
///
/// This module contains a class `Solution` with a method `merge`.
/// The `merge` method takes in four parameters:
/// - `nums1`: A list of integers representing the first sorted array.
/// - `m`: An integer representing the number of initial elements in `nums1`.
/// - `nums2`: A list of integers representing the second sorted array.
/// - `n`: An integer representing the number of elements in `nums2`.
///
/// The method does not return anything. It modifies `nums1` in-place such that it contains the elements of both `nums1` and `nums2` in sorted order.
///
/// Error Handling:
/// - This method assumes that the input parameters are well-formed, i.e., `nums1` and `nums2` are lists of integers, and `m` and `n` are integers.
/// - If the input parameters are not well-formed, the behavior of the method is undefined.

class Solution {
  /// Merges two sorted arrays in-place.
  ///
  /// This method uses three pointers to traverse `nums1` and `nums2` from the end and places the larger element at the end of `nums1`.
  ///
  /// @param nums1 The first sorted array.
  /// @param m The number of initial elements in `nums1`.
  /// @param nums2 The second sorted array.
  /// @param n The number of elements in `nums2`.
  void merge(List<int> nums1, int m, List<int> nums2, int n) {
    // Initialize the pointers at the end of `nums1`, `nums2`, and the merged array.
    int i = m - 1;
    int j = n - 1;
    int k = m + n - 1;

    // While there are elements in both `nums1` and `nums2`, place the larger element at the end of the merged array.
    while (i >= 0 && j >= 0) {
      if (nums1[i] > nums2[j]) {
        nums1[k--] = nums1[i--];
      } else {
        nums1[k--] = nums2[j--];
      }
    }

    // If there are remaining elements in `nums2`, place them at the beginning of the merged array.
    while (j >= 0) {
      nums1[k--] = nums2[j--];
    }
  }
}
